pkg/api/handlers/compat/volumes.go

```diff
@@ -0,0 +1,263 @@
+package compat
+
+import (
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/containers/libpod/libpod"
+	"github.com/containers/libpod/libpod/define"
+	"github.com/containers/libpod/pkg/api/handlers/utils"
+	"github.com/containers/libpod/pkg/domain/filters"
+	"github.com/containers/libpod/pkg/domain/infra/abi/parse"
+	"github.com/gorilla/schema"
+	"github.com/pkg/errors"
+
+	docker_api_types "github.com/docker/docker/api/types"
+	docker_api_types_volume "github.com/docker/docker/api/types/volume"
+)
+
+// swagger:response DockerVolumeList
+type swagDockerVolumeListResponse struct {
+	// in:body
+	Body struct {
+		docker_api_types_volume.VolumeListOKBody
+	}
+}
+
+func ListVolumes(w http.ResponseWriter, r *http.Request) {
+	var (
+		decoder = r.Context().Value("decoder").(*schema.Decoder)
+		runtime = r.Context().Value("runtime").(*libpod.Runtime)
+	)
+	query := struct {
+		Filters map[string][]string `schema:"filters"`
+	}{
+		// override any golang type defaults
+	}
+
+	if err := decoder.Decode(&query, r.URL.Query()); err != nil {
+		utils.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest,
+			errors.Wrapf(err, "Failed to parse parameters for %s", r.URL.String()))
+		return
+	}
+
+	// XXX: This will reject `dangling` in the `filters` since we don't yet
+	// understand it there. This manifests as a 500 response complaining about
+	// the "invalid" volume filter.
+	// XXX: This will also accept the `opts` filter which isn't included in the
+	// Docker engine v1.40 API spec.
+	volumeFilters, err := filters.GenerateVolumeFilters(query.Filters)
+	if err != nil {
+		utils.InternalServerError(w, err)
+		return
+	}
+
+	vols, err := runtime.Volumes(volumeFilters...)
+	if err != nil {
+		utils.InternalServerError(w, err)
+		return
+	}
+	volumeConfigs := make([]*docker_api_types.Volume, 0, len(vols))
+	for _, v := range vols {
+		config := docker_api_types.Volume{
+			Name:       v.Name(),
+			Driver:     v.Driver(),
+			Mountpoint: v.MountPoint(),
+			CreatedAt:  v.CreatedTime().Format(time.RFC3339),
+			Labels:     v.Labels(),
+			Scope:      v.Scope(),
+			Options:    v.Options(),
+		}
+		volumeConfigs = append(volumeConfigs, &config)
+	}
+	response := docker_api_types_volume.VolumeListOKBody{
+		Volumes:  volumeConfigs,
+		Warnings: []string{},
+	}
+	utils.WriteResponse(w, http.StatusOK, response)
+}
+
+// swagger:model DockerVolumeCreate
+type DockerVolumeCreate docker_api_types_volume.VolumeCreateBody
+
+// This response definition is used for both the create and inspect endpoints
+// swagger:response DockerVolumeInfoResponse
+type swagDockerVolumeInfoResponse struct {
```

Comment:
The linter complains about these swagger response defs being dead code. It's right, and I don't really know if swagger can even discover them here (IDK how it works and haven't checked out how to run the build/doc steps locally). I put them inline here since I made the choice to put the `DockerVolumeCreate` typedef in here. These have a kind of logical home in `api/server/swagger.go`, but I figured for now I'd keep everything in one place and just point out that it needs some thought from someone with more insight than me.

Reply:
I think this needs to move into `pkg/entities/` with the other structs, and probably be public - those two things should resolve the issue

Reply:
As above, these live in `pkg/entities` now